portal/src/graphql/adminapi/AddUserScreen.tsx

```diff
@@ -1,140 +1,329 @@
 import { useTextField } from "../../hook/useInput";
 import { PasswordPolicyConfig, PortalAPIAppConfig } from "../../types";
+import {
+  defaultFormatErrorMessageList,
+  Violation,
+} from "../../util/validation";
+import { parseError } from "../../util/error";
 
 import styles from "./AddUserScreen.module.scss";
 
+type LoginIDKey = "username" | "email" | "phone";
+function isLoginIDKey(value?: string): value is LoginIDKey {
+  return ["username", "email", "phone"].includes(value ?? "");
+}
+
 interface AddUserContentProps {
   appConfig: PortalAPIAppConfig | null;
 }
 
 interface AddUserScreenState {
-  name: string;
+  selectedLoginIdKey?: LoginIDKey;
   username: string;
   email: string;
   phone: string;
   password: string;
 }
 
-interface FieldVisible {
-  username: boolean;
-  email: boolean;
-  phone: boolean;
-  password: boolean;
+interface LoginIdIdentityOptionProps {
+  option?: IChoiceGroupOption;
+  renderTextField?: () => React.ReactNode;
 }
 
-interface AddUserViolation {
-  type: "OneOfLoginIdType" | "PasswordInvalid";
+const loginIdLocaleKey: Record<LoginIDKey, string> = {
+  username: "login-id-key.username",
+  email: "login-id-key.email",
+  phone: "login-id-key.phone",
+};
+
+function determineIsPasswordNeeded(
+  appConfig: PortalAPIAppConfig | null,
+  loginIdKeySelected: LoginIDKey | undefined
+) {
+  if (loginIdKeySelected == null) {
+    return false;
+  }
+  const primaryAuthenticators =
+    appConfig?.authentication?.primary_authenticators ?? [];
+  // password is first one, all login ID types need password
+  if (primaryAuthenticators[0] === "password") {
+    return true;
+  }
+  // password is second one, require password if user choose username
+  // only id is username -> need password
+  if (deepEqual(["oob_otp", "password"], primaryAuthenticators)) {
+    return loginIdKeySelected === "username";
+  }
+  return false;
 }
 
-function checkOnlyOneOfLoginIdFieldFilled(screenState: AddUserScreenState) {
-  return (
-    [screenState.username, screenState.email, screenState.phone].filter(
-      (fieldValue) => fieldValue.length > 0
-    ).length === 1
-  );
+function getLoginIdKeyOptions(appConfig: PortalAPIAppConfig | null) {
+  const primaryAuthenticators =
+    appConfig?.authentication?.primary_authenticators ?? [];
+
+  // need password authenticator in order to use username to login
+  const usernameAllowed = primaryAuthenticators.includes("password");
+
+  const loginIdKeys = appConfig?.identity?.login_id?.keys ?? [];
+  const loginIdKeyOptions = new Set<LoginIDKey>();
+  for (const key of loginIdKeys) {
+    switch (key.type) {
+      case "username": {
+        if (usernameAllowed) {
+          loginIdKeyOptions.add("username");
+        }
+        break;
+      }
+      case "email":
+        loginIdKeyOptions.add("email");
+        break;
+      case "phone":
+        loginIdKeyOptions.add("phone");
+        break;
+      default:
+        break;
+    }
+  }
+  return Array.from(loginIdKeyOptions);
```

Comment:
We should also check whether login ID identity is enabled (i.e. `login_id` is contained in `authentication.identities`).

Reply:
How should we handle the case where login ID identity is not enabled? create user requires loginIDIdentities, I think we need a simple empty state in case where user cannot be added? @kiootic

Reply:
I see there is an existing empty state when user cannot be added. If I understood correctly, returning an empty array from the function should trigger that state?

Reply:
Yes, that's correct. However it is a text only empty state which does not look good... I added `TODO` in the corresponding line. I have returned empty array as there is no possible login ID identity option if login ID identity itself is disabled.